docs/architecture/adr-041-in-place-store-migrations.md

```diff
@@ -0,0 +1,174 @@
+
+### Module-Specific Migration Scripts
+
+For each consensus-breaking change introduced by the module, a migration script from ConsensusVersion `N` to version `N+1` MUST be registered in the `Configurator` using its newly-added `RegisterMigration` method. All modules receive a reference to the configurator in their `RegisterServices` method on `AppModule`, and this is where the migration scripts should be registered.
+
+```go
+func (am AppModule) RegisterServices(cfg module.Configurator) {
+    // --snip--
+    cfg.RegisterMigration(types.ModuleName, 1, func(ctx sdk.Context) error {
+        // Perform in-place store migrations from ConsensusVersion 1 to 2.
+    })
+     cfg.RegisterMigration(types.ModuleName, 2, func(ctx sdk.Context) error {
+        // Perform in-place store migrations from ConsensusVersion 2 to 3.
+    })
+    // etc.
+}
+```
+
+For example, if the current ConsensusVersion of a module is `N` , then `N-1` migration scripts MUST be registered in the configurator.
+
+In the SDK, the migration scripts are handled by each module's keeper, because the keeper holds the `sdk.StoreKey` used to perform in-place store migrations. A `MigrationKeeper` interface is implemented by each keeper:
+
+```go
+// MigrationKeeper is an interface that the keeper implements for handling
+// in-place store migrations.
+type MigrationKeeper interface {
+    // Migrate1 migrates the store from version 1 to 2.
+    Migrate1(ctx sdk.Context) error
+
+    // ...Add more MigrateN scripts here.
+}
+```
+
+Since migration scripts manipulate legacy code, they should live inside the `legacy/` folder of each module, and be called by the keeper's implementation of `MigrationKeeper`.
+
+```go
+func (keeper BankKeeper) Migrate1(ctx sdk.Context) error {
+    return v042bank.MigrateStore(ctx, keeper.storeKey) // v042bank is package `x/bank/legacy/v042`.
+}
+```
+
+Each module's migration scripts are specific to the module's store evolutions, and are not described in this ADR. An example of x/bank store key migrations following the introduction of ADR-028 length-prefixed addresses can be seen [here](https://github.com/cosmos/cosmos-sdk/blob/ef8dabcf0f2ecaf26db1c6c6d5922e9399458bb3/x/bank/legacy/v042/store.go#L15).
+
+### Tracking Module Versions in `x/upgrade`
+
+We introduce a new prefix store in `x/upgrade`'s store. This store will track each module's current version, it can be modelized as a `map[string]uint64` of module name to module ConsensusVersion, and will be used when running the migrations (see next section for details). The key prefix used is `0x1`, and the key/value format is:
+
+```
+0x1 | {module_name_bytes} => LittleEndian(module_consensus_version)
+```
+
+We add a new parameter `ModuleManager` `x/upgrade`'s `NewKeeper` constructor, where ModuleManager is:
```

Comment:
Why `ModuleManager` goes to the `x/upgrade.Keeper` constructor? I thought that the `x/upgrade.QueryServer`  will provide that functionality.  In that case we would simply query the `x/upgrade` and as for old versions.

Reply:
To query the current versions from loaded modules

Reply:
I updated this a little bit: we unfortunately cannot add it in the constructor, because of a definition cycle in app.go: UpgradeKeeper needs ModuleManager which needs UpgradeKeeper. The way proposed here is to add a `SetVersionManager` setter on UpgradeKeeper.